### DISCUSSION TITLE
"Install from VSIX" using local file system

### DISCUSSION THREAD

**Discussion by liechtir** [coder/code-server#4897] [Q&A]
Hi 

I run vscode server sucessfully on my docker container, can access the UI and install extensions like GitLens or Extension Pack for Java.

However, I cannot find the Redhat server connector and other wildfly/jboss related extensions at all in the marketplace.
When I donwload the VSIX file and try to install manually, I receive a "No Servers" error.
How can I use wildfly from VSCode when vscode runs on a remote server? VScode-server and wildfly are on the same server under the same user.

Thanks

**Comment by jsjoeio** [comment]
> I receive a "No Servers" error.

I have never seen that before. Very strange. 

> How can I use wildfly from VSCode when vscode runs on a remote server

I'm not sure what wildfly. Can you post steps to reproduce or try this locally?

**Reply by liechtir** [reply]
I tried to install the extention locally in a VSCode that runs on my windows directly. There it works like a charm, I can install the extension and create a new server (Wildfly 24 in this case), the server is downloaded and can be started.

However, when I run vscode-sever on my docker container I cannot even see the redhat extension in the marketplace.

Steps to reproduce (like they are executed in my dockerfile):

1. create vscode server folders and  install vscode-server:
mkdir vscodeData && mkdir vscodeExtensions && curl -fsSL https://code-server.dev/install.sh | sh -s
2. make sure the docker container users (named wildly) has permissions to those folders:
sudo chown wildfly:1000 ~/vscodeExtensions ~/vscodeData
3. boot up vscode-server
code-server --auth none --user-data-dir ~/vscodeData --extensions-dir ~/vscodeExtensions --bind-addr 0.0.0.0:1337 &
4.  open browser and go to localhost:1337
5. VSCode UI shows up
6. go to extensions, search for 'Server Connector' or any other redhat related extensions, such as 'Remote Server Protocol UI', none of them is found.
7. Download the extensions manually from https://marketplace.visualstudio.com/items?itemName=redhat.vscode-server-connector or https://marketplace.visualstudio.com/items?itemName=redhat.vscode-rsp-ui
8. Try to install them manually using the downlaoded file, receive the "no Server" error.

thanks!

**Reply by jsjoeio** [reply]
> go to extensions, search for 'Server Connector' or any other redhat related extensions, such as 'Remote Server Protocol UI', none of them is found.

code-server uses the Open VSX registry and I'm not seeing 'Server Connector' on there. I suggest you reach out asking them to add it.

I downloaded the VSIX file and then installed in the UI in code-server and it worked:

```logs
[10:09:25] Installing extension: redhat.vscode-server-connector
[10:09:30] Extracted extension to /Users/jp/.local/share/code-server/extensions/.d9b1f054-4f73-4cf3-8ade-103357f732ea: redhat.vscode-server-connector
[10:09:31] Renamed to /Users/jp/.local/share/code-server/extensions/redhat.vscode-server-connector-0.25.2
[10:09:31] Extracting completed. redhat.vscode-server-connector
[10:09:31] Extension installed successfully: redhat.vscode-server-connector
```

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/3806031/155573037-ec86ffb7-7eff-4858-98c3-6aa9d23c9bc0.png">

Can you give that a try?

And can you post the logs where you're seeing "no Server" error?

**Reply by liechtir** [reply]
Yes this works when you have the application installed locally like you do.
When you install code-server on a remote server and then use your browser to connect to the vs code client, it does not work.

**Comment by liechtir** [comment]
I get the 'No Server' error in the right lower corner of the UI, cannot find any details on the UI. There is no further information printed on the console where code-server was started.
![image](https://user-images.githubusercontent.com/1678080/155987440-4a5f0a56-d675-49f7-952c-f73524009834.png)

In any of the logs I found, there is nothing that tells me anything:
![image](https://user-images.githubusercontent.com/1678080/155987676-cab7e5f3-a043-439a-a91a-1aa4d40660ac.png)

Any special place you want me to look for?

**Reply by jsjoeio** [reply]
No, these screenshots are very helpful! I have never seen that "No Servers" error before in the UI 🤔

I found where this code lives: https://github.com/microsoft/vscode/blob/7e0df7ed8ff74b00da41d7bd20e7a6ceab3d6603/src/vs/workbench/services/extensionManagement/common/extensionManagementService.ts#L224-L225

That sounds like it can't get the manifest from the vsix file? 🤔 I'm not sure why that would happen. Maybe something is wrong with the `this.extensionManagementServerService.remoteExtensionManagementServer`? 

Recap of your environment:
- docker container
- accessing remotely 
- Wildfly 24 server?

Is there anything else that might be related?

Have you tested other extensions?

**Comment by liechtir** [comment]
wow that helped! :-)
So it cannot read the manifest because the upload ot the extension from my local pc to the remote server worked, but the file could most likely not be stored due to permissions.
So when I choose 'show local' here, it does not work:
![image](https://user-images.githubusercontent.com/1678080/156122429-4e4d0a05-6426-479c-bc7d-b6b947c8aad0.png)

However, if I upload the file to the server using 'docker cp redhat.vscode-server-connector-0.25.2.vsix containerId:/some/writeable/path' and then choose the file from this tree, installation works perfectly:
![image](https://user-images.githubusercontent.com/1678080/156122608-a487ee98-f83c-46d7-b3f2-56c43f668970.png)

So we can workaround the issue. How can we fix it?
I am booting up the server using 'code-server --auth none --user-data-dir ~/vscodeData --extensions-dir ~/vscodeExtensions' , is there any other directory the server needs permission to?`I installed the server using the install.sh from github using the linux user that runs the server.

**Reply by jsjoeio** [reply]
Wow! I am happy that lead to a solution - go team!

> I am booting up the server using 'code-server --auth none --user-data-dir ~/vscodeData --extensions-dir ~/vscodeExtensions' , is there any other directory the server needs permission to?`I installed the server using the install.sh from github using the linux user that runs the server.

Great question! Let me ask @code-asher - summary: is there any other directory that needs permissions specifically extensions and extension manifests?

**Reply by liechtir** [reply]
it's not related to the redhat extension, any extension that is not available on the https://open-vsx.org/ martketplace, so that I download it from the microsoft marketplace, cannot be installed by uploading using the UI from VScode but I need to upload it using file transfer to the remote server and then choose to install by browsing the local file system of the remote server.
All mention the 'No Server' error.

**Reply by liechtir** [reply]
@jsjoeio, any luck getting in touch with @code-asher ?

**Reply by code-asher** [reply]
I think this is an upstream bug. You end up with a `file:///extension-name.vsix` URI which tries to use the local extension service but of course there is no local extension service in the web.  Even if there was that path is not a valid one.

The proper resolution is probably to add some code that uploads the extension to the server and then installs it.  I looked for upstream issues but I have not found one, assuming we can replicate in Codespaces we could file there.

**Reply by code-asher** [reply]
Another workaround is to drag the extension onto the explorer sidebar so it gets uploaded then install from there.

**Reply by liechtir** [reply]
> I think this is an upstream bug. You end up with a `file:///extension-name.vsix` URI which tries to use the local extension service but of course there is no local extension service in the web. Even if there was that path is not a valid one.
> 
> The proper resolution is probably to add some code that uploads the extension to the server and then installs it. I looked for upstream issues but I have not found one, assuming we can replicate in Codespaces we could file there.

So what one could do is to simply not show the button 'Show Local' when the UI connects to a remote instance.

**Reply by code-asher** [reply]
Agreed, this sounds like a reasonable solution.
